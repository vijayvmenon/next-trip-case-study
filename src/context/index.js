import { useContext } from 'react';
import AppContext, { Context } from './context';

export default AppContext;
export { Context };
export const useMyContext = () => useContext(Context);
