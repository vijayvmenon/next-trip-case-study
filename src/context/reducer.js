import constants from '../constants';
import ENTranslations from '../locales/en.json';
import { localStorageExists } from '../utils';

const localStorageLanguage =
  localStorageExists && localStorage.getItem('LANGUAGE');
const initialState = {
  tableExpanded: false,
  hideHeader: false,
  snackbar: { value: '', text: '' },
  language: localStorageLanguage || 'en',
  translations: localStorageLanguage
    ? constants.LANGUAGE_MAPPING[localStorageLanguage]
    : ENTranslations,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TABLE_EXPAND':
      return { ...state, tableExpanded: true };
    case 'TABLE_COLLAPSE':
      return { ...state, tableExpanded: false };
    case 'SET_SNACKBAR':
      return { ...state, snackbar: { value: action.value, text: action.text } };
    case 'HIDE_HEADER':
      return { ...state, hideHeader: true };
    case 'SHOW_HEADER':
      return { ...state, hideHeader: false };
    case 'SET_LANGUAGE':
      return {
        ...state,
        language: action.value,
        translations: constants.LANGUAGE_MAPPING[action.value],
      };
    default:
      return { ...state };
  }
}

export { initialState };
