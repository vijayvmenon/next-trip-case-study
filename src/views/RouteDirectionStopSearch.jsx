/* eslint-disable */
import React, { useEffect, useState } from 'react';
import { createUseStyles } from 'react-jss';
import { useLocation } from 'wouter';
import Select from '../components/Select';
import constants from '../constants';
import { httpGet, urlBuilder, checkIfUrlHasCorrectParams } from '../utils';
import TableWrapper from './TableWrapper';

const { ROUTE, DIRECTION, STOP, http, SELECT_FIELDS } = constants;

const useStyles = createUseStyles({
  root: {
    marginTop: 24,
    width: '60%',
  },
  tablePadding: {
    padding: '32px 60px',
  },
});

export default function RouteDirectionStopSearch() {
  const classes = useStyles();
  const [dropdowns, setDropDowns] = useState(SELECT_FIELDS);
  const [error, setError] = useState(false);
  const [location, setLocation] = useLocation();
  const [selectProperties, setSelectProperties] = useState([
    { id: ROUTE, param: null, selected: '', active: false },
    { id: DIRECTION, param: null, selected: '', active: false },
    { id: STOP, param: null, selected: '', active: false },
  ]);
  const [tableData, setTableData] = useState(null);
  const [intrvel, setIntrvl] = useState('initial');

  const onSearch = (value, source, idField) => {
    setSelectProperties(
      selectProperties.map((val) => {
        // logic to reset selected value on route or direction dropdown change accordingly. Also sets the url param
        let selectedValue = val.selected;
        if (val.id === source) {
          selectedValue = value[idField];
        }
        if (source === ROUTE && (val.id === DIRECTION || val.id === STOP)) {
          selectedValue = '';
        }
        if (source === DIRECTION && val.id === STOP) {
          selectedValue = '';
        }
        return {
          ...val,
          param: val.id === source ? value[idField] : val.param,
          selected: selectedValue,
          active: val.id === source ? true : false,
        };
      }),
    );
  };

  // Load routes on Initial Load
  useEffect(() => {
    getData(http.ROUTE, ROUTE);
    console.log(location, checkIfUrlHasCorrectParams(location));
    const urlToLoad = checkIfUrlHasCorrectParams(location);
    if (urlToLoad) {
      getTableData('on-search', urlToLoad);
    }
  }, []);

  // Moving the data fetching logic for dropdowns and table to the useeffect once selectProperties state is changed.
  // Add logic to not run this on initial load as well
  useEffect(() => {
    const notInitialLoad = selectProperties.find((val) => val.param !== null);
    if (notInitialLoad) {
      const { id: activeDropDown, param: activeParam } = selectProperties.find(
        (val) => val.active,
      );
      setError(false);
      // Reset table on change of route or directions dropdown . Also clear Interval if set. Does not run on initial load
      if (tableData !== null) {
        setTableData(null);
        setIntrvl('dropdown-changed');
      }
      if (activeDropDown === ROUTE) {
        getData(`${http.DIRECTIONS}/${activeParam}`, DIRECTION); //get directions data again based on route selected
      } else if (activeDropDown === DIRECTION) {
        const routeParam = selectProperties.find(
          (val) => val.id === ROUTE,
        ).param;
        getData(`${http.STOPS}/${routeParam}/${activeParam}`, STOP); //get stops data again based on direction selected
      } else {
        getTableData('on-search'); //get table data when stop is changed
      }
    }
  }, [selectProperties]);

  function getTableData(interval, passedUrl) {
    /* set tableData to null only if a new search is made. 
    This is to handle the useEffect run for clearing interval in <Table/> component */
    interval === 'on-search' && setTableData(null);
    setIntrvl(interval);
    const tableUrl = passedUrl || urlBuilder(selectProperties);
    setLocation(`/byRoute/${tableUrl}`);
    getData(tableUrl, 'table');
  }

  function setDropdownData(data, source) {
    setDropDowns(
      dropdowns.map((val) => ({
        ...val,
        /* set the dropdowndata based on the source passed from getDropDownData function. When the source is DIRECTION,
       i.e when Routes dropdown is changed, then we need to hide the stops dropdown. Below logic handles that */
        options:
          val.labelSelected === source
            ? data
            : val.labelSelected === STOP && source === DIRECTION
            ? undefined
            : val.options,
      })),
    );
  }

  async function getData(param, source) {
    const [response, error] = await httpGet(param);
    //  console.log(response, error);
    if (response) {
      if (source === 'table') {
        setTableData(response);
      } else {
        setDropdownData(response, source);
      }
    }
    if (error) {
      setIntrvl('error');
      setError(true);
    }
  }

  function getSelectedValue(val) {
    const result = selectProperties.find(
      (innerVal) => innerVal.id === val.labelSelected,
    ).selected;
    if (result === '') {
      return val.label;
    } else {
      return result;
    }
  }

  return (
    <>
      <form className={classes.root} action="">
        {dropdowns
          .filter((val) => val.options)
          .map((val) => (
            <Select
              key={val.label}
              label={val.label}
              onSearch={onSearch}
              options={val.options}
              labelSelected={val.labelSelected}
              idField={val.idField}
              valueField={val.valueField}
              selectedValue={getSelectedValue(val)}
            />
          ))}
      </form>
      <TableWrapper
        tableData={tableData}
        rowField="departures"
        tableHeaderField="stops"
        interval={intrvel}
        getTableData={getTableData}
        setIntrvl={(val) => setIntrvl(val)}
      />
    </>
  );
}
