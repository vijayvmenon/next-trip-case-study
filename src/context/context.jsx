import React, { useReducer, createContext } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';

export const Context = createContext();

export default function AppContext({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
}

AppContext.propTypes = {
  children: PropTypes.objectOf(PropTypes.any).isRequired,
};
