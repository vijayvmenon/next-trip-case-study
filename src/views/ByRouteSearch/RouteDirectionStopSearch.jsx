import React, { useEffect, useState, useContext } from 'react';
import { createUseStyles } from 'react-jss';
import { useLocation } from 'wouter';
import { Context } from '../../context';
import Select from '../../components/UI/Select';
import constants from '../../constants';
import { httpGet, urlBuilder, checkIfUrlIsInByRouteFormat } from '../../utils';
import TableWrapper from '../TableWrapper/TableWrapper';
import Loader from '../../components/UI/Loader';
import Snackbar from '../../components/UI/Snackbar';

const useStyles = createUseStyles({
  root: {
    marginTop: 24,
    width: '100%',
    position: 'relative',
    maxWidth: 570,
  },
  tablePadding: {
    padding: '32px 60px',
  },
});

export default function RouteDirectionStopSearch() {
  const classes = useStyles();
  const {
    dispatch,
    state: {
      snackbar: { value: snackbarType = '', text = '' } = {},
      translations = {},
    } = {},
  } = useContext(Context);

  const { http, SELECT_FIELDS } = constants;

  const [dropdowns, setDropDowns] = useState(SELECT_FIELDS);
  const [location, setLocation] = useLocation();
  const [loader, setLoader] = useState(false);
  const [tableLoader, setTableLoader] = useState(false);
  const [routeOnReload, setRouteOnReload] = useState('');
  const [selectProperties, setSelectProperties] = useState([
    { id: 'ROUTE', param: null, selected: '', active: false },
    { id: 'DIRECTION', param: null, selected: '', active: false },
    { id: 'STOP', param: null, selected: '', active: false },
  ]);
  const [tableData, setTableData] = useState(null);
  const [intrvel, setIntrvl] = useState('initial');
  const [backButtonClicked, setBackButtonClicked] = useState(false);

  const onSearch = (value, source, idField) => {
    // Reset params in url when user searches through the dropdowns again
    setRouteOnReload('');
    setSelectProperties(
      selectProperties.map((val) => {
        // logic to reset selected value on route or direction dropdown change accordingly. Also sets the url param
        let selectedValue = val.selected;
        if (val.id === source) {
          selectedValue = value[idField];
        }
        if (
          source === 'ROUTE' &&
          (val.id === 'DIRECTION' || val.id === 'STOP')
        ) {
          selectedValue = '';
        }
        if (source === 'DIRECTION' && val.id === 'STOP') {
          selectedValue = '';
        }
        return {
          ...val,
          param: val.id === source ? value[idField] : val.param,
          selected: selectedValue,
          active: val.id === source,
        };
      }),
    );
  };

  function setDropdownData(data, source) {
    function getOptions(val) {
      if (val.labelSelected === source) {
        return data;
      }
      if (val.labelSelected === 'STOP' && source === 'DIRECTION') {
        return undefined;
      }
      return val.options;
    }
    setDropDowns(
      dropdowns.map((val) => ({
        ...val,
        /* set the dropdowndata based on the source passed from getDropDownData function. When the source is DIRECTION,
       i.e when Routes dropdown is changed, then we need to hide the stops dropdown. Below logic handles that */
        options: getOptions(val),
      })),
    );
  }

  async function getData(param, source) {
    if (source === 'table') {
      setTableLoader(true);
    } else {
      setLoader(true);
    }

    const [fetchResponse, fetchError] = await httpGet(param);
    if (fetchResponse) {
      if (source === 'table') {
        setTableData(fetchResponse);
      } else {
        setDropdownData(fetchResponse, source);
      }
    }
    if (fetchError) {
      setIntrvl('error');
      if (source !== 'table') {
        setDropdownData([], source); // set next dropdown to empty options when error, but not for table as its not required
      }
      // Open snackbar based on warning or error from the API call
      if (fetchError === 'warning') {
        dispatch({
          type: 'SET_SNACKBAR',
          value: 'warning',
          text: translations.NO_DEPARTURE_ROUTE,
        });
      } else {
        dispatch({
          type: 'SET_SNACKBAR',
          value: 'error',
          text: `${translations.ERROR_FETCHING}${
            source === 'table' ? translations.DEPARTURES : source
          }. ${translations.PLEASE_TRY_AGAIN}`,
        });
      }
    }
    setLoader(false);
    setTableLoader(false);
  }

  function getTableData(interval, passedUrl) {
    /* set tableData to null only if a new search is made. 
    This is to handle the useEffect run for clearing interval in <Table/> component */
    if (interval === 'on-search') {
      setTableData(null);
    }
    setIntrvl(interval);
    // call the departures API on interval from tableWrapper, on dropdown change or on initial load
    const tableUrl = passedUrl || routeOnReload || urlBuilder(selectProperties);
    //  setLocation(`/byRoute/${tableUrl}`);
    getData(tableUrl, 'table');
  }

  /* Load departures (if url has required parameters) on Initial Load . This useEffect also loads the departures data when
  user searches something by changing the stops dropdown. Since location is passed in the array of dependencies and 
  we are setting location in line 198, this useEffect will run and get the TableData. 
  It is written like this to handle situation when the user clicks browser back button */
  useEffect(() => {
    async function fetchData() {
      dispatch({ type: 'SHOW_HEADER' }); // show Header component again when redirecting from a page not found
      const activeProperty = selectProperties.find((val) => val.active);
      const urlToLoad = checkIfUrlIsInByRouteFormat(location);

      // Load routes only on initial load
      if (!activeProperty) {
        await getData(http.ROUTE, 'ROUTE');
      }
      if (urlToLoad) {
        setRouteOnReload(urlToLoad);
        getTableData('on-search', urlToLoad);
      } else {
        /* Handle situation where user presses browser back button and comes to the root /byRoute page. We need to reset
      the interval if any are set */
        setTableData(null);
        setIntrvl('reset');
        // Show warning snackbar if the URL parameters are not valid and only when not in the rool url
        if (location !== '/byRoute') {
          dispatch({
            type: 'SET_SNACKBAR',
            value: 'warning',
            text: translations.NOT_VALID_SEARCH,
          });
        }
      }
    }
    fetchData();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch, location]);

  // Moving the data fetching logic for dropdowns and table to the useeffect once selectProperties state is changed.
  // Add logic to not run this on initial load as well
  useEffect(() => {
    const notInitialLoad = selectProperties.find((val) => val.param !== null);
    if (notInitialLoad) {
      const { id: activeDropDown, param: activeParam } = selectProperties.find(
        (val) => val.active,
      );
      // Reset table on change of route or directions dropdown . Also clear Interval if set. Does not run on initial load
      if (tableData !== null) {
        setTableData(null);
        setIntrvl('dropdown-changed');
      }
      if (activeDropDown === 'ROUTE') {
        getData(`${http.DIRECTIONS}/${activeParam}`, 'DIRECTION'); // get directions data again based on route selected
      } else if (activeDropDown === 'DIRECTION') {
        const routeParam = selectProperties.find(
          (val) => val.id === 'ROUTE',
        ).param;
        getData(`${http.STOPS}/${routeParam}/${activeParam}`, 'STOP'); // get stops data again based on direction selected
      } else {
        const tableUrl = urlBuilder(selectProperties);
        setLocation(`/byRoute/${tableUrl}`);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectProperties]);

  function getSelectedValue(val) {
    const result = selectProperties.find(
      (innerVal) => innerVal.id === val.labelSelected,
    ).selected;
    if (result === '') {
      return val.label;
    }
    return result;
  }

  /* This useEffect is to handle browser back button click situation. The 'popstate" event listener will run 
   on initial load and based on the backButtonclicked state, when the user presses back button and the URL is a
   valid URL to show departures, it will reload the table based on the URL params , but hide th other dropdowns 
   except route dropdown, in order to give the same experience as a browser refresh  */
  useEffect(() => {
    window.addEventListener('popstate', async () => {
      // set the state only if its not set to true. This is an optimization
      if (!backButtonClicked) {
        setBackButtonClicked(true);
      }
    });
    if (backButtonClicked) {
      setDropDowns(
        dropdowns.map((val) => ({
          ...val,
          options: val.labelSelected !== 'ROUTE' ? undefined : val.options,
        })),
      );
      setBackButtonClicked(false);
    }
    return () => window.removeEventListener('popstate', {});
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [backButtonClicked]);

  return (
    <>
      <Snackbar type={snackbarType} text={text} />
      <form className={classes.root} action="">
        {loader && <Loader top="20%" />}
        {dropdowns
          .filter((val) => val.options !== undefined)
          .map((val) => (
            <Select
              key={val.label}
              label={val.label}
              onSearch={onSearch}
              options={val.options}
              labelSelected={val.labelSelected}
              idField={val.idField}
              valueField={val.valueField}
              selectedValue={getSelectedValue(val)}
            />
          ))}
      </form>
      <TableWrapper
        tableData={tableData}
        rowField="departures"
        tableHeaderField="stops"
        interval={intrvel}
        getTableData={getTableData}
        setIntrvl={(val) => setIntrvl(val)}
        loader={tableLoader}
      />
    </>
  );
}
