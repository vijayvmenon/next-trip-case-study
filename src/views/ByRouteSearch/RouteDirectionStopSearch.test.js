import { render, fireEvent, screen } from '@testing-library/react';
import RouteDirectionStopSearch from './RouteDirectionStopSearch';
import renderer from 'react-test-renderer';
import { Context } from '../../context';

const dispatch = jest.fn(); // mock dispatch

describe('Component:RouteDirectionStopSearch.jsx', () => {
  test('RouteDirectionStopSearch.jsx renders correctly', () => {
    const routeDirectionStop = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <RouteDirectionStopSearch />
        </Context.Provider>,
      )
      .toJSON();
    expect(routeDirectionStop).toMatchSnapshot();
  });
});
