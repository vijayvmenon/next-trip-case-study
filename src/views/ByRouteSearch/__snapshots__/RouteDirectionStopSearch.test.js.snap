// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:RouteDirectionStopSearch.jsx RouteDirectionStopSearch.jsx renders correctly 1`] = `
Array [
  "",
  <form
    action=""
    className="root-0-2-1"
  >
    <div
      className="root-0-2-11"
    >
      <select
        className="select-0-2-18 notSelected-0-2-15"
        data-testid="custom-select-route_id"
        onChange={[Function]}
        value="SELECT_ROUTE"
      >
        <option
          data-testid="option-default-route_id"
          disabled={true}
          hidden={false}
          value="SELECT_ROUTE"
        />
      </select>
    </div>
  </form>,
]
`;
